sdk/metric/instrument.go

```diff
@@ -24,10 +24,157 @@ import (
+	// InstrumentKindSyncUpDownCounter identifies a group of instruments that
+	// record increasing and decreasing values synchronously with the code path
+	// they are measuring.
+	InstrumentKindSyncUpDownCounter
+	// InstrumentKindSyncHistogram identifies a group of instruments that
+	// record a distribution of values synchronously with the code path they
+	// are measuring.
+	InstrumentKindSyncHistogram
+	// InstrumentKindAsyncCounter identifies a group of instruments that record
+	// increasing values in an asynchronous callback.
+	InstrumentKindAsyncCounter
+	// InstrumentKindAsyncUpDownCounter identifies a group of instruments that
+	// record increasing and decreasing values in an asynchronous callback.
+	InstrumentKindAsyncUpDownCounter
+	// InstrumentKindAsyncGauge identifies a group of instruments that record
+	// current values in an asynchronous callback.
+	InstrumentKindAsyncGauge
+)
+
+type nonComparable [0]func() // nolint: unused  // This is indeed used.
+
+// Instrument describes properties an instrument is created with.
+type Instrument struct {
+	// Name is the human-readable identifier of the instrument.
+	Name string
+	// Description describes the purpose of the instrument.
+	Description string
+	// Kind defines the functional group of the instrument.
+	Kind InstrumentKind
+	// Unit is the unit of measurement recorded by the instrument.
+	Unit unit.Unit
+	// Scope identifies the instrumentation that created the instrument.
+	Scope instrumentation.Scope
+
+	// Ensure forward compatibility if non-comparable fields need to be added.
+	nonComparable // nolint: unused
+}
+
+// mask returns a copy of p with all non-zero-value fields of m replacing the
+// fields of the returned copy.
+func (p Instrument) mask(m Instrument) Instrument {
+	if m.Name != "" {
+		p.Name = m.Name
+	}
+	if m.Description != "" {
+		p.Description = m.Description
+	}
+	if m.Kind != zeroInstrumentKind {
+		p.Kind = m.Kind
+	}
+	if m.Unit != zeroUnit {
+		p.Unit = m.Unit
+	}
+	if m.Scope.Name != "" {
+		p.Scope.Name = m.Scope.Name
+	}
+	if m.Scope.Version != "" {
+		p.Scope.Version = m.Scope.Version
+	}
+	if m.Scope.SchemaURL != "" {
+		p.Scope.SchemaURL = m.Scope.SchemaURL
+	}
+	return p
+}
+
+// empty returns if all fields of p are their zero-value.
+func (p Instrument) empty() bool {
+	return p.Name == "" &&
+		p.Description == "" &&
+		p.Kind == zeroInstrumentKind &&
+		p.Unit == zeroUnit &&
+		p.Scope == zeroScope
+}
+
+// matches returns if all the non-zero-value fields of o match the
+// corresponding fields of p.
+//
+// If o is empty true is returned.
+func (p Instrument) matches(o Instrument) bool {
+	return p.matchesName(o) &&
+		p.matchesDescription(o) &&
+		p.matchesKind(o) &&
+		p.matchesUnit(o) &&
+		p.matchesScope(o)
+}
+
+// matchesName returns true if the Name field of o is a non-zero-value and
+// equals the Name field of p, otherwise false.
+func (p Instrument) matchesName(o Instrument) bool {
+	return p.Name == "" || p.Name == o.Name
```

Comment:
```suggestion
	return o.Name == "" || p.Name == o.Name
```

Should the first check here be against `o` instead of `p`?  That would make the comment about an empty `o` above correct.

Reply:
The comment was incorrect. It should only evaluate equality if `p` is not `""`. If it is `""` it should "match" everything and return `true.

I've updated the comment. PTAL.